### PR TITLE
#8855 - OpenAI new components

### LINKED ISSUE #8855: [Components] openai
# openai

#### Prompt
Use the List fine-tuning jobs API to poll for new items after the last-processed job.

This source has no props besides the OpenAI account.

Please handle pagination. You'll see that there's an after string that can be used to retrieve items since the last-processed item. Please also use service.db to maintain state about the most-recent item ID, so you can request items only after that ID.

On the deploy hook, the source should emit the last 10 fine-tuning jobs. It should then keep track of the after string in the last item and save that in service.db state. In the run method, you should only fetch fine-tuning jobs since the after string.
#### URLs
- https://raw.githubusercontent.com/openai/openai-openapi/master/openapi.yaml
#### Prompt
Use the List files API to poll for new items after the last-processed job.

The source should have one optional prop: purpose.
#### URLs
- https://raw.githubusercontent.com/openai/openai-openapi/master/openapi.yaml

#### Prompt
Require these props:

- model — One of the available TTS models: `tts-1` or `tts-1-hd`
- input — The text to generate audio for. The maximum length is 4096 characters.
- voice — The voice to use when generating the audio. Supported voices are alloy, echo, fable, onyx, nova, and shimmer.

Optional props:

- response_format — The format to audio in. Supported formats are mp3, opus, aac, and flac.
- speed — The speed of the generated audio. Select a value from 0.25 to 4.0. 1.0 is the default.
#### URLs
- https://raw.githubusercontent.com/openai/openai-openapi/master/openapi.yaml

#### Prompt
Require the following props:

- model — one of gpt-3.5-turbo-1006 (recommended), babbage-002, davinci-002, gpt-4-0613
- training_file — See upload file for how to upload a file. Your dataset must be formatted as a JSONL file. Additionally, you must upload your file with the purpose fine-tune. See the fine-tuning guide for more details. The prop should be a string and ask the user to provide a reference to a file on `/tmp` that they wrote in a previous step.

The rest of the params to this endpoint are optional.
#### URLs
- https://raw.githubusercontent.com/openai/openai-openapi/master/openapi.yaml
#### Prompt
The Upload File API accepts two required props:

- file: a File object in a multipart-form request (example: curl -F file="@filename"). The prop should be a string and ask the user to provide a reference to a file on `/tmp` that they wrote in a previous step.
- purpose: Use "fine-tune" for Fine-tuning and "assistants" for Assistants and Messages. You should pass this in the form request (example: curl -F purpose="fine-tune")
#### URLs
- https://raw.githubusercontent.com/openai/openai-openapi/master/openapi.yaml

### PR DESCRIPTION
**Pull request by GTFalcao**
### <samp>🤖 Generated by Copilot at 5644be9</samp>

This pull request adds new action and source components for the OpenAI integration, and improves the existing ones. It adds the ability to create fine-tuning jobs and speech from text, and to list and upload files with more options. It also refactors some of the common logic and constants into separate files, and updates the dependencies and the version of the OpenAI package.

### <samp>🤖 Generated by Copilot at 5644be9</samp>

> _We are the masters of OpenAI_
> _We create, we fine-tune, we generate_
> _We stream the files and the jobs to the sky_
> _We use the form-data to upload our fate_

## WHY

Closes #8855 

### <samp>🤖 Generated by Copilot at 5644be9</samp>

*  Add new action components to create speech and fine-tuning jobs in OpenAI ([link](https://github.com/PipedreamHQ/pipedream/pull/8906/files?diff=unified&w=0#diff-0ba66581f0ab83233be79e56e25588b64cde61dfd2054a7dc2ef02ea0e754fa7R1-R57), [link](https://github.com/PipedreamHQ/pipedream/pull/8906/files?diff=unified&w=0#diff-7cb1dee1c6e202ce925f431bb00a2c8b7de77a76d0dd9e9c28bf0e84e0e89424R1-R57))
*  Update upload file action component to use form-data and fs modules for handling file content as a stream ([link](https://github.com/PipedreamHQ/pipedream/pull/8906/files?diff=unified&w=0#diff-c0d6f2df0798d593f5488207183e0711a8e953fb74e9c8fc18dce796babc2e35L2-R43), [link](https://github.com/PipedreamHQ/pipedream/pull/8906/files?diff=unified&w=0#diff-89cff885df98c3ad74c1ebaed8db649f7a9ad42fe10fbfdbb40adb4a1a756e56L244-R311), [link](https://github.com/PipedreamHQ/pipedream/pull/8906/files?diff=unified&w=0#diff-89cff885df98c3ad74c1ebaed8db649f7a9ad42fe10fbfdbb40adb4a1a756e56L594-R660), [link](https://github.com/PipedreamHQ/pipedream/pull/8906/files?diff=unified&w=0#diff-d53db1f6f1c2069172138aaf09b971481bfa0777e04f58f6b56b34c2cd4bc171L21-R22), [link](https://github.com/PipedreamHQ/pipedream/pull/8906/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR4125), [link](https://github.com/PipedreamHQ/pipedream/pull/8906/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR4132))
*  Add optional property to purpose input of list files action component to filter files by purpose or list all files ([link](https://github.com/PipedreamHQ/pipedream/pull/8906/files?diff=unified&w=0#diff-45c4f1660d1a7d707d8ef3d23bea78f2f627edcf7861b16fd331809b380155b2R16))
*  Update version of list files action component from 0.0.1 to 0.0.2 ([link](https://github.com/PipedreamHQ/pipedream/pull/8906/files?diff=unified&w=0#diff-45c4f1660d1a7d707d8ef3d23bea78f2f627edcf7861b16fd331809b380155b2L7-R7))
*  Add new constant file that exports fine-tuning model options for user selection ([link](https://github.com/PipedreamHQ/pipedream/pull/8906/files?diff=unified&w=0#diff-f59513fef9696ed20136b62020954b04e44be6edacc488edb78cd01f7f99d030R1-R23))
*  Import fine-tuning model options from constant file into `openai.app.mjs` file ([link](https://github.com/PipedreamHQ/pipedream/pull/8906/files?diff=unified&w=0#diff-89cff885df98c3ad74c1ebaed8db649f7a9ad42fe10fbfdbb40adb4a1a756e56R2))
*  Reorder and add more props for create speech and create fine-tuning job actions in `openai.app.mjs` file, using propDefinition property to reference common props ([link](https://github.com/PipedreamHQ/pipedream/pull/8906/files?diff=unified&w=0#diff-89cff885df98c3ad74c1ebaed8db649f7a9ad42fe10fbfdbb40adb4a1a756e56L206-R277))
*  Modify _makeRequest method in `openai.app.mjs` file to merge headers from args parameter with common headers, instead of overriding them ([link](https://github.com/PipedreamHQ/pipedream/pull/8906/files?diff=unified&w=0#diff-89cff885df98c3ad74c1ebaed8db649f7a9ad42fe10fbfdbb40adb4a1a756e56L244-R311))
*  Add new methods to `openai.app.mjs` file for interacting with OpenAI API endpoints for list fine-tuning jobs, create speech, and create fine-tuning job ([link](https://github.com/PipedreamHQ/pipedream/pull/8906/files?diff=unified&w=0#diff-89cff885df98c3ad74c1ebaed8db649f7a9ad42fe10fbfdbb40adb4a1a756e56R682-R701))
*  Update version of openai package from 0.2.0 to 0.3.0 and change main property to point to `openai.app.mjs` file instead of app folder ([link](https://github.com/PipedreamHQ/pipedream/pull/8906/files?diff=unified&w=0#diff-d53db1f6f1c2069172138aaf09b971481bfa0777e04f58f6b56b34c2cd4bc171L3-R5))
*  Add new common file that exports base source component for OpenAI, defining common props, methods, and hooks for fetching and emitting data from OpenAI API ([link](https://github.com/PipedreamHQ/pipedream/pull/8906/files?diff=unified&w=0#diff-559cd475b2afff8321086dd2443f8b15e82d58501a09d9f6e3fbffe8f43f655dR1-R51))
*  Add new source components that emit events when new files or fine-tuning jobs are created in OpenAI, extending common source component and defining specific props and methods ([link](https://github.com/PipedreamHQ/pipedream/pull/8906/files?diff=unified&w=0#diff-ef6e21166a24f2f26e5ec7face344f4f3f88199e12011b18d0a46466a4204d46R1-R37), [link](https://github.com/PipedreamHQ/pipedream/pull/8906/files?diff=unified&w=0#diff-5c8119adc4080450157ed160d9698aa35a542ad5f5c44c771e53a05e26d26055R1-R23))

### CODE REVIEW

File: components/openai/actions/create-speech/create-speech.mjs [hunk 1]
```diff
@@ -0,0 +1,57 @@
+import openai from "../../openai.app.mjs";
+
+export default {
+  key: "openai-create-speech",
+  name: "Create Speech",
+  description: "Generates audio from the input text. [See the documentation](https://platform.openai.com/docs/api-reference/audio/createSpeech)",
+  version: "0.0.2",
```
Comment (luancazarine) [inline components/openai/actions/create-speech/create-speech.mjs:7]:
```suggestion
  version: "0.0.1",
```

File: components/openai/actions/list-files/list-files.mjs [hunk 2]
```diff
@@ -4,7 +4,7 @@ export default {
   key: "openai-list-files",
   name: "List Files",
   description: "Returns a list of files that belong to the user's organization. [See the documentation](https://platform.openai.com/docs/api-reference/files/list)",
-  version: "0.0.1",
+  version: "0.0.3",
```
Comment (luancazarine) [inline components/openai/actions/list-files/list-files.mjs:7]:
```suggestion
  version: "0.0.2",
```

File: components/openai/actions/retrieve-run-step/retrieve-run-step.mjs [hunk 3]
```diff
@@ -4,7 +4,7 @@ export default {
   key: "openai-retrieve-run-step",
   name: "Retrieve Run Step",
   description: "Retrieve a specific run step in a thread. [See the documentation](https://platform.openai.com/docs/api-reference)",
-  version: "0.0.1",
+  version: "0.0.3",
```
Comment (luancazarine) [inline components/openai/actions/retrieve-run-step/retrieve-run-step.mjs:7]:
```suggestion
  version: "0.0.2",
```

File: components/openai/actions/upload-file/upload-file.mjs [hunk 4]
```diff
@@ -1,41 +1,46 @@
...
+import FormData from "form-data";
+import fs from "fs";
 
 export default {
   key: "openai-upload-file",
   name: "Upload File",
-  description: "Upload a file that can be used across various endpoints/features. The size of individual files can be a maximum of 512mb. [See the documentation](https://beta.openai.com/docs/guides/files)",
-  version: "0.0.4",
+  description: "Upload a file that can be used across various endpoints/features. The size of individual files can be a maximum of 512mb. [See the documentation](https://platform.openai.com/docs/api-reference/files/create)",
+  version: "0.0.6",
```
Comment (luancazarine) [inline components/openai/actions/upload-file/upload-file.mjs:9]:
```suggestion
  version: "0.0.5",
```

File: components/openai/sources/new-file-created/new-file-created.mjs [hunk 5]
```diff
@@ -0,0 +1,38 @@
+import openai from "../../openai.app.mjs";
+import common from "../common.mjs";
+
+export default {
+  ...common,
+  key: "openai-new-file-created",
+  name: "New File Created",
+  description: "Emit new event when a new file is created in OpenAI. [See the documentation](https://platform.openai.com/docs/api-reference/files/list)",
+  version: "0.0.2",
```
Comment (luancazarine) [inline components/openai/sources/new-file-created/new-file-created.mjs:9]:
```suggestion
  version: "0.01",
```

File: components/openai/sources/new-fine-tuning-job-created/new-fine-tuning-job-created.mjs [hunk 6]
```diff
@@ -0,0 +1,24 @@
+import common from "../common.mjs";
+
+export default {
+  ...common,
+  key: "openai-new-fine-tuning-job-created",
+  name: "New Fine Tuning Job Created",
+  description: "Emit new event when a new fine-tuning job is created in OpenAI. [See the documentation](https://platform.openai.com/docs/api-reference/fine-tuning/list)",
+  version: "0.0.2",
```
Comment (luancazarine) [inline components/openai/sources/new-fine-tuning-job-created/new-fine-tuning-job-created.mjs:8]:
```suggestion
  version: "0.0.1",
```

**Review by luancazarine — CHANGES_REQUESTED** [review]
Hi @GTFalcao, thanks for your contribution! I just added some comments.

**Review by luancazarine — DISMISSED** [review]
LGTM!

File: components/openai/actions/create-fine-tuning-job/create-fine-tuning-job.mjs [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    trainingFile: {
+      propDefinition: [
+        openai,
+        "trainingFile",
+      ],
+    },
+    hyperParameters: {
+      type: "object",
+      label: "Hyperparameters",
+      description: "The hyperparameters used for the fine-tuning job. [See details on the documentation.](https://platform.openai.com/docs/api-reference/fine-tuning/create#fine-tuning-create-hyperparameters)",
```
Comment (dylburger) [inline components/openai/actions/create-fine-tuning-job/create-fine-tuning-job.mjs:26]: "See details on" -> "See details in" and move the period at the end outside of the link.

File: components/openai/actions/create-fine-tuning-job/create-fine-tuning-job.mjs [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    suffix: {
+      type: "string",
+      label: "Suffix",
+      description: "A string of up to 18 characters that will be added to your fine-tuned model name.",
+      optional: true,
+    },
+    validationFile: {
+      type: "string",
+      label: "Validation File",
+      description: "The ID of an uploaded file that contains validation data. [See details on the documentation.](https://platform.openai.com/docs/api-reference/fine-tuning/create#fine-tuning-create-validation_file)",
```
Comment (dylburger) [inline components/openai/actions/create-fine-tuning-job/create-fine-tuning-job.mjs:38]: Same feedback as https://github.com/PipedreamHQ/pipedream/pull/8906/files#r1390631526

File: components/openai/actions/create-speech/create-speech.mjs [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+    speed: {
+      propDefinition: [
+        openai,
+        "speed",
+      ],
+    },
+    outputFile: {
+      type: "string",
+      label: "Output Filename",
+      description: "The filename of the output audio file that will be written to the `/tmp` folder, e.g. `myFile.mp3`",
```
Comment (dylburger) [inline components/openai/actions/create-speech/create-speech.mjs:45]: e.g. `myFile.mp3` -> `/tmp/myFile.mp3`

File: components/openai/openai.app.mjs [hunk 10]
```diff
@@ -203,16 +209,76 @@ export default {
...
+      description: "One of the available [TTS models](https://platform.openai.com/docs/models/tts).",
+      options: [
+        "tts-1",
+        "tts-1-hd",
+      ],
+    },
+    fineTuningModel: {
+      type: "string",
+      label: "Fine Tuning Model",
+      description: "The name of the model to fine-tune. [See the supported models.](https://platform.openai.com/docs/guides/fine-tuning/what-models-can-be-fine-tuned)",
```
Comment (dylburger) [inline components/openai/openai.app.mjs:238]: Move the period out of the link

**Review by dylburger — APPROVED** [review]

**Comment by andrewjschuang** [comment]
@GTFalcao I created a PR to merge with this branch: https://github.com/PipedreamHQ/pipedream/pull/8944
Could you please give a review (and feel free to merge)

**Review by andrewjschuang — APPROVED** [review]

**Review by andrewjschuang — APPROVED** [review]

**Pull request merged**